lib/guest_installation_and_configuration_base.pm

```diff
@@ -0,0 +1,1465 @@
+    assert_script_run("touch $_guest_installation_log && chmod 777 $_guest_installation_log");
+    if (script_run("set -o pipefail;$self->{virt_install_command_line_dryrun} 2>&1 | tee -a $_guest_installation_dryrun_log", timeout => 600 / get_var('TIMEOUT_SCALE', 1)) ne 0) {
+        record_info("Guest $self->{guest_name} installation dry run failed", "The virt-install command used is $self->{virt_install_command_line_dryrun}");
+        $self->record_guest_installation_result('FAILED');
+        script_run("set +o pipefail");
+        save_screenshot;
+        return $self;
+    }
+    script_run("set +o pipefail");
+    save_screenshot;
+    record_info("Guest $self->{guest_name} installation dry run succeeded", "Going to install by using $self->{virt_install_command_line}");
+    type_string("screen -t $self->{guest_name} -L -Logfile $_guest_installation_log $self->{virt_install_command_line}\n", timeout => 600 / get_var('TIMEOUT_SCALE', 1));
+    record_info("Guest $self->{guest_name} installation started", "The virt-install command line is $self->{virt_install_command_line}");
+    return $self;
+}
+
+#Get guest installation screen process information and store it in [guest_installation_session] which is in the form of 3401.pts-1.vh017.
+sub get_guest_installation_session {
+    my $self = shift;
+
+    $self->reveal_myself;
+    if ($self->{guest_installation_session} ne '') {
+        record_info("Guest $self->{guest_name} installation screen process info had already been known", "$self->{guest_name} $self->{guest_installation_session}");
+        return $self;
+    }
+    my $installation_tty = script_output("tty | awk -F\"/\" 'BEGIN { OFS=\"-\" } {print \$3,\$4}\'", proceed_on_failure => 1);
+    my $installation_pid = script_output("pgrep -i -f \"SCREEN -t $self->{guest_name}\"",            proceed_on_failure => 1);
+    $self->{guest_installation_session} = ($installation_pid eq '' ? '' : $installation_pid . ".$installation_tty." . $self->{host_name});
+    record_info("Guest $self->{guest_name} installation screen process info", "$self->{guest_name} $self->{guest_installation_session}");
+    return $self;
+}
+
+#Kill all guest installation screen processes stored in [guest_installation_session] after test finishes.
+sub terminate_guest_installation_session {
+    my $self = shift;
+
+    $self->reveal_myself;
+    if ($self->{guest_installation_session} ne '') {
+        script_run("screen -X -S $self->{guest_installation_session} kill");
+        record_info("Guest $self->{guest_name} installation screen process $self->{guest_installation_session} has already been terminated now", "Installation already passed or failed");
+    }
+    else {
+        record_info("Guest $self->{guest_name} has no associated installation screen process to be terminated", "This looks weird");
+    }
+    return $self;
+}
+
+#Get dynamic allocated guest ip address using nmap scan and store it in [guest_ipaddr].
+sub get_guest_ipaddr {
```

Comment:
Seems similar nmap way to get guest ip addr, like https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12334.

Let's not continue this discussion in this PR. We can arrange a discussion meeting for exceptions of "code duplication as little as possible" rule after sle15sp3 GM. We can have it merged just as it is now. Regarding next step, if after the meeting, this nmap way of getting guest address duplication meets exception rules, we will keep it. Otherwise, we can make new small PR to fix the duplication within the 3 files.

Reply:
This is not "any kind of duplicaion" here in this pull request. If this is going to be somewhere is openQA lib file, I have no objection to do this in principle, but might not be able to do this at the moment right away.